File: capabilities/auth/targets/flutter/files/lib/src/auth/credential_store.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../core/storage/secure_storage.dart';
import '../core/storage/secure_storage_provider.dart';

/// Key of the refresh credential in the platform keystore. Scoped and stable —
/// [SecureStorage] rejects anything else.
const refreshCredentialKey = 'auth.refresh_token';

/// Where the refresh credential lives, as a replaceable seam.
///
/// The interface exists for the same reason Angular's `CredentialStore` and
/// React Native's `SecureStorage` do: the client depends on a contract, never on
/// a platform API. The three runtimes differ by implementation, never by
/// architecture — which is what makes parity measurable instead of declarative.
///
/// Unlike a browser, a phone does offer a protected store. The default
/// implementation therefore *persists*, through the platform keystore: on mobile,
/// "the most protected store the platform actually offers" is a real thing, and
/// keeping the credential in memory would throw away protection the platform
/// grants for free (ADR-076).
abstract interface class CredentialStore {
  /// Returns the stored credential, or `null` when there is none.
  Future<String?> read();

  /// Stores [value], replacing any previous credential.
  Future<void> write(String value);

  /// Removes the credential. A no-op when there is none.
  Future<void> clear();
}

/// Default store: the mobile keystore, reached through the baseline
/// secure-storage contract rather than through a plugin API.
final class SecureCredentialStore implements CredentialStore {
  const SecureCredentialStore(this._storage);

  final SecureStorage _storage;

  @override
  Future<String?> read() => _storage.get(refreshCredentialKey);

  @override
  Future<void> write(String value) => _storage.set(refreshCredentialKey, value);

  @override
  Future<void> clear() => _storage.remove(refreshCredentialKey);
}

/// Volatile store, for tests and for a deployment that deliberately refuses to
/// persist anything. It makes no promise it cannot keep.
final class InMemoryCredentialStore implements CredentialStore {
  String? _value;

  @override
  Future<String?> read() async => _value;

  @override
  Future<void> write(String value) async {
    _value = value;
  }

  @override
  Future<void> clear() async {
    _value = null;
  }
}

final credentialStoreProvider = Provider<CredentialStore>(
  (ref) => SecureCredentialStore(ref.watch(secureStorageProvider)),
);

File: capabilities/auth/targets/angular/files/src/app/features/auth/credential-store.ts
/**
 * Where the refresh credential lives, as a replaceable seam (ADR-075).
 *
 * The interface exists for the same reason React Native's `SecureStorage` does:
 * the client depends on the contract, never on a platform API. The two runtimes
 * differ by implementation, never by architecture.
 *
 * A browser offers a static SPA no protected store — anything a script can write,
 * a script can read. The default implementation therefore keeps the credential in
 * memory: it cannot be exfiltrated by injected script from a previous page load,
 * and it disappears with the tab. The cost is explicit: a full reload signs the
 * user out.
 *
 * `localStorage` is deliberately absent and must stay absent: it would persist a
 * thirty-day credential in a store readable by any injected script.
 */
export interface CredentialStore {
  /** Returns the stored credential, or `null` when there is none. */
  read(): string | null;
  /** Stores `value`, replacing any previous credential. */
  write(value: string): void;
  /** Removes the credential. A no-op when there is none. */
  clear(): void;
}

/**
 * Default store: process memory only.
 *
 * Chosen as the default because it is the only browser option that makes no
 * false promise. A deployment that serves the API and the application same-site
 * should prefer an HttpOnly refresh cookie issued by the authority — the client
 * then holds nothing at all — which is why this is a seam and not a hard-coded
 * mechanism.
 */
export class InMemoryCredentialStore implements CredentialStore {
  #value: string | null = null;

  read(): string | null {
    return this.#value;
  }

  write(value: string): void {
    this.#value = value;
  }

  clear(): void {
    this.#value = null;
  }
}
